nupic/research/frameworks/pytorch/imagenet/mixins/save_final_checkpoint.py

```diff
@@ -0,0 +1,133 @@
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import os
+from operator import itemgetter
+from pathlib import Path
+from shutil import copytree
+
+
+class SaveFinalCheckpoint(object):
+    """
+    Experiment mixin to save a copy of the final checkpoint from `logdir` upon
+    `stop_experiment`.
+
+    :param config:
+
+        - copy_checkpoint_dir: (optional) name of directory to save a copy of the final
+                               checkpoint. The string may contain:
+                                    * {name} - name of experiment; taken from config
+                                    * {wandb_project} - taken from wandb_args
+                                    * {trainable_id} - inferred from logdir
+
+    Example
+    ```
+    config["name"] = "my_experiment"
+    config["logdir"] = ".../ImagenetTrainable_...-45xrhk_mc2"
+    config["wandb_args"] = dict(..., project="my_project")
+    config["copy_checkpoint_dir"] = "/mnt/.../{wandb_project}/{name}/{trainable_id}/"
+
+    ... 200 epochs of training
+
+    >> checkpoint saved to "/mnt/.../my_project/my_experiment/45xrhk_mc2/checkpoint_200"
+    ```
+    """
+
+    def setup_experiment(self, config):
+        """Setup experiment and get desired path for copied checkpoint."""
+        super().setup_experiment(config)
+
+        assert self.logdir
+        assert "copy_checkpoint_dir" in config
+
+        # Set default formatting.
+        self.copy_checkpoint_dir = config["copy_checkpoint_dir"]
+        self.name = config.get("name", "unknown_name")
+        self.wandb_project = "unknown_project"
+        self.trainable_id = "unknown_id"
+
+        # Retrieve project name if any - this may or may not be used.
+        if "{wandb_project}" in self.copy_checkpoint_dir:
+            wandb_args = config.get("wandb_args", {})
+            self.wandb_project = wandb_args.get("project", "unknown_project")
+
+        # Retrieve run_id - this may or may not be used.
+        if "{trainable_id}" in self.copy_checkpoint_dir:
+            self.trainable_id = self.logdir.split("-")[-1]  # e.g. 45xrhk_mc2
+
+        # Set up to directory that will contain the copied checkpoint.
+        self.setup_copy_dir()
+
+    def setup_copy_dir(self):
```

Comment:
If this is executed at every run of this mixin, maybe we should add it to the eo list in `get_execution_order`?

Reply:
Fair point. Added in newest commit below.